src/st/st0/2DAC8.c

```diff
@@ -1127,105 +1187,109 @@ void func_801AF774(Entity* self) {
         if (self->ext.et_801AF774.unk8E > 0) {
             func_801AF6D0();
         }
-        var_t0 = (u16)self->ext.et_801AF774.unk8C >> 8;
-        if (var_t0 >= 0x81) {
-            var_t0 = 0x80;
+        rgb = self->ext.et_801AF774.unk8C >> 8;
+        if (rgb > 0x80) {
+            rgb = 0x80;
         }
-        var_t0 = (u8)var_t0 >> 3;
-        prim = self->ext.prim;
+        rgb >>= 3;
+        prim = self->ext.et_801AF774.prim1;
         xBase = 0x40;
         yBase = 0x70;
-        temp_a3 = &D_801BEB64;
-        var_a2_3 = &D_801813E8;
+        points = *D_801BEB64;
+        indexes = D_801813E8;
 
         for (i = 0; i < 47; i++) {
-            prim->x0 = add(xBase, temp_a3[*var_a2_3 * 4 + 1]);
-            prim->y0 = add(yBase, temp_a3[*var_a2_3 * 4 + 3]);
-            var_a2_3++;
-            prim->x1 = add(xBase, temp_a3[*var_a2_3 * 4 + 1]);
-            prim->y1 = add(yBase, temp_a3[*var_a2_3 * 4 + 3]);
-            var_a2_3++;
-            prim->x2 = add(xBase, temp_a3[*var_a2_3 * 4 + 1]);
-            prim->y2 = add(yBase, temp_a3[*var_a2_3 * 4 + 3]);
-            var_a2_3++;
-            prim->x3 = add(xBase, temp_a3[*var_a2_3 * 4 + 1]);
-            prim->y3 = add(yBase, temp_a3[*var_a2_3 * 4 + 3]);
-            var_a2_3++;
-            prim->b0 = var_t0;
-            prim->g0 = var_t0;
-            prim->r0 = var_t0;
+            // The way this is written is horrendous
+            // but required for match on PSP
+            prim->x0 = (((*indexes * 2)[points]) >> 16) + xBase;
+            prim->y0 = (((*indexes++ * 2)[points + 1]) >> 16) + yBase;
```

Comment:
This had a really tricky `addu` register swap on PSP. This was the only syntax I could find to resolve it.

If anyone has any good ideas please let me know
https://decomp.me/scratch/jQNRR